### PR TITLE
Enforced immutable fields using CEL rules

### LINKED ISSUE #4128: Enforce immutable fields using CEL rules
### What problem are you facing?

Many Crossplane API fields are immutable - you can't change them after they're first set. Typically today when a field is immutable we'll let you update it, and will just silently ignore the update. This is obviously a terrible user experience.

We've wanted to be able to enforce immutability at the schema level for a long time, but until recently doing so was pretty involved. You essentially had to write a webhook (e.g. https://github.com/crossplane/crossplane/issues/727) for every type to enforce immutability.

Now with CEL and kubebuilder it's possible to do this much more conveniently with comment markers - see https://kubernetes.io/blog/2022/09/29/enforce-immutability-using-cel/. Once these comment markers are added any API server that supports CEL will reject updates to immutable fields.

### How could Crossplane help solve your problem?

We should use CEL-based immutability.

In many cases we've already added `// +immutable` comment markers on fields we know are immutable. These comment markers don't do anything - they're just placeholders and should be replaced with CEL ones.

### PR DESCRIPTION
**Pull request by NeerajNagure**
### Description of your changes
Added CEL based comment markers to enforce immutability on every field that was previously marked as // +immutable

Fixes #4128 

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] ~Added or updated unit tests.~
- [x]  Added or updated e2e tests.
- [ ] ~Linked a PR or a [docs tracking issue] to [document this change].~
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR.~

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet

### CODE REVIEW

File: apis/apiextensions/v1/composition_revision_types.go [hunk 1]
```diff
@@ -38,7 +38,9 @@ const (
 type CompositionRevisionSpec struct {
 	// CompositeTypeRef specifies the type of composite resource that this
 	// composition is compatible with.
-	// +immutable
+	// +kubebuilder:validation:Optional
```
Comment (negz) [inline apis/apiextensions/v1/composition_revision_types.go:41]: Why mark this field optional? It isn't.

I don't think any of the fields you've added the `Optional` comment marker to are optional.

File: apis/apiextensions/v1/composition_revision_types.go [hunk 2]
```diff
@@ -38,7 +38,9 @@ const (
...
+	// +kubebuilder:validation:XValidation:rule="!has(oldSelf.value) || has(self.value)", message="Value is required once set"
```
Comment (negz) [inline apis/apiextensions/v1/composition_revision_types.go:42]: Does this actually make the field immutable? It looks like it prevents it from being removed, but doesn't prevent it from being changed.

**Review by negz — COMMENTED** [review]
Did you run `make generate`? It doesn't seem like the CRDs were updated with the result of the new comment marker.

Also, I'm pretty confident we have more immutable fields than this. There's probably some without `// +immutable` markers.

File: apis/apiextensions/v1/composition_revision_types.go (hunk 1 above)
Comment (NeerajNagure) [reply apis/apiextensions/v1/composition_revision_types.go]: @negz I have fixed those fields that are not supposed to be optional

File: apis/apiextensions/v1/composition_revision_types.go (hunk 2 above)
Comment (NeerajNagure) [reply apis/apiextensions/v1/composition_revision_types.go:42]: @negz Actually I have already added `+kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"` just below that line which will look into making the field immutable.

**Comment by NeerajNagure** [comment]
@negz Also can I please know how can I find those fields that are supposed to be immutable but are not marked with //+immutable ?

File: apis/apiextensions/v1/composition_revision_types.go (hunk 2 above)
Comment (negz) [reply apis/apiextensions/v1/composition_revision_types.go:42]: Do we need both rules, `self == oldSelf` and also `!has(oldSelf.value) || has(self.value)`?

Also, does the latter rule actually prevent a field being removed? The logic appears to be "if old self doesn't have the value, or self has the value". Should it be `&&` not `||`?

File: apis/apiextensions/v1/xrd_types.go [hunk 3]
```diff
@@ -30,12 +30,16 @@ type CompositeResourceDefinitionSpec struct {
 	// Group specifies the API group of the defined composite resource.
 	// Composite resources are served under `/apis/<group>/...`. Must match the
 	// name of the XRD (in the form `<names.plural>.<group>`).
-	// +immutable
+	// +kubebuilder:validation:Required
```
Comment (negz) [inline apis/apiextensions/v1/xrd_types.go:33]: Is there a reason to mark fields as required or optional in this PR? Unless enforcing immutability requires this change, I would suggest keeping the change focused only on immutability.

**Comment by negz** [comment]
> Also can I please know how can I find those fields that are supposed to be immutable but are not marked with //+immutable ?

@NeerajNagure Unfortunately there's no easy way to do this. You'd need to be familiar enough with Crossplane's APIs to know which fields can't change. I don't think this needs to block this PR though, if we can come up with a pattern that works someone who is familiar with the APIs can copy it to the other fields that need it.

File: apis/apiextensions/v1/xrd_types.go (hunk 3 above)
Comment (NeerajNagure) [reply apis/apiextensions/v1/xrd_types.go:33]: @negz its not necessary to mark fields as required or optional to enforce immutability. I will update this.

Comment (NeerajNagure) [reply apis/apiextensions/v1/xrd_types.go:33]: @negz when the kubebuilder validation is set to required then we enforce immutability upon object creation time and when its set to optional then we enforce immutability after first modification i.e. immutability after its first write. So I think its required for enforcing immutability.

**Comment by NeerajNagure** [comment]
@negz can you please review this?

File: apis/apiextensions/v1/composition_revision_types.go (hunk 2 above)
Comment (NeerajNagure) [reply apis/apiextensions/v1/composition_revision_types.go]: @negz I have removed that rule now and we only check `self==oldself`

File: apis/apiextensions/v1/xrd_types.go (hunk 3 above)
Comment (NeerajNagure) [reply apis/apiextensions/v1/xrd_types.go]: @negz I have removed the kubebuilder validation

**Comment by NeerajNagure** [comment]
@negz I have made the changes as suggested by you

**Comment by negz** [comment]
Thanks @NeerajNagure. Could you please add some details on how you've tested this change? Adding an E2E test would be ideal.

**Comment by NeerajNagure** [comment]
@negz I have added 2 e2e tests that check the if its possible to update immutable fields or not

**Review by negz — APPROVED** [review]
Thanks @NeerajNagure!

**Comment by MisterMX** [comment]
@negz is it necessary for `spec.enforcedCompositionRef` to be an immutable field? It feels very countintuitive to have admin enforcement settings fixed on a specific value once set. Especially if you have made a mistake or want to enforce an upgrade. There also no other way since you cannot delete an XRD without affecting existing claims ...

**Pull request merged**